### PR TITLE
Add coverage

### PR DESCRIPTION
**Pull request by Fristi**
Resolving #87 

- [x] Build
- [x] Publish to codecov, @julienrf would you like that as coverage?
- [x] Rebase

### CODE REVIEW

**Comment by Fristi** [comment]
Some notes

I've tried running `sbt coverage +test coverageReport coverageAggregate`, but that fails due to
`[error] /home/travis/build/julienrf/endpoints/xhr/client/src/main/scala/endpoints/xhr/Assets.scala:68: Found a dangling UndefinedParam at Position(file:/home/travis/build/julienrf/endpoints/xhr/client/src/main/scala/endpoints/xhr/Assets.scala,70,12). This is likely due to a bad interaction between a macro or a compiler plugin and the Scala.js compiler plugin. If you hit this, please let us know.`

So I switched to a staged build which, builds all three supported Scala versions in parallel in the first test stage. After that's done we run only the coverage for the JVM projects. We disable coverage for all JS projects. We might want to turn that on later, if the issues are fixed (compile error) and the aggregation of results is correct (because if you run JVM+JS coverage on the same piece of code, it shouldn't add up).

Fails on this piece https://travis-ci.org/julienrf/endpoints/jobs/361147097:

```
java.lang.RuntimeException: No source root found for '/home/travis/build/julienrf/endpoints/documentation/examples/cqrs/public-server/target/assets/BootstrapDigests.scala' (source roots: '/home/travis/build/julienrf/endpoints/documentation/examples/cqrs/public-server/src/main/scala/')
	at scoverage.report.BaseReportWriter.relativeSource(BaseReportWriter.scala:28)
	at scoverage.report.BaseReportWriter.relativeSource(BaseReportWriter.scala:16)
	at scoverage.report.CoberturaXmlWriter.klass(CoberturaXmlWriter.scala:42)
	at scoverage.report.CoberturaXmlWriter$$anonfun$pack$1.apply(CoberturaXmlWriter.scala:66)
	at scoverage.report.CoberturaXmlWriter$$anonfun$pack$1.apply(CoberturaXmlWriter.scala:66)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at scoverage.report.CoberturaXmlWriter.pack(CoberturaXmlWriter.scala:66)
	at scoverage.report.CoberturaXmlWriter$$anonfun$xml$3.apply(CoberturaXmlWriter.scala:90)
	at scoverage.report.CoberturaXmlWriter$$anonfun$xml$3.apply(CoberturaXmlWriter.scala:90)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.immutable.List.foreach(List.scala:318)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.AbstractTraversable.map(Traversable.scala:105)
	at scoverage.report.CoberturaXmlWriter.xml(CoberturaXmlWriter.scala:90)
	at scoverage.report.CoberturaXmlWriter.write(CoberturaXmlWriter.scala:20)
	at scoverage.ScoverageSbtPlugin$.scoverage$ScoverageSbtPlugin$$writeReports(ScoverageSbtPlugin.scala:177)
	at scoverage.ScoverageSbtPlugin$$anonfun$coverageReport0$1.apply(ScoverageSbtPlugin.scala:114)
	at scoverage.ScoverageSbtPlugin$$anonfun$coverageReport0$1.apply(ScoverageSbtPlugin.scala:105)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by Fristi** [comment]
It builds now. Now we need to setup codecov or something similiar

**Comment by julienrf** [comment]
Hi, that’s a great initiative!

I’ve enabled codecov for this repository. IIUC you just have to add the `codecov` command to the TravisCI script.

**Comment by julienrf** [comment]
The documentation projects do contain some tests (e.g. https://github.com/julienrf/endpoints/blob/master/documentation/examples/cqrs/infra/src/test/scala/cqrs/Test.scala). It’s better if we can make them work with scoverage.

**Comment by Fristi** [comment]
> I’ve enabled codecov for this repository. IIUC you just have to add the codecov command to the TravisCI script.

Did you set the `CODECOV_TOKEN` in travis? :-)

> The documentation projects do contain some tests (e.g. https://github.com/julienrf/endpoints/blob/master/documentation/examples/cqrs/infra/src/test/scala/cqrs/Test.scala). It’s better if we can make them work with scoverage.

All the tests are run in the first stage, even the ScalaJS ones :-)

**Comment by julienrf** [comment]
@Fristi They say that the tokens are not necessary for public repos. And there is this `codecov` command: https://github.com/codecov/example-python/blob/master/.travis.yml#L15

**Comment by julienrf** [comment]
> All the tests are run in the first stage, even the ScalaJS ones :-)

Sure, what I meant was that it would be nice if those tests were taken into account by scoverage.

**Comment by Fristi** [comment]
Should be good to go! In #89 you could enable the coverage for the example projects again

File: documentation/build.sbt [hunk 1]
```diff
@@ -31,6 +31,7 @@ val apiDoc =
   project.in(file("api-doc"))
     .settings(noPublishSettings ++ `scala 2.11` ++ unidocSettings: _*)
     .settings(
+      coverageEnabled := false,
```
Comment (julienrf) [inline documentation/build.sbt]: Would you mind removing these since I’ve merged #89?

**Comment by Fristi** [comment]
Incorporated your changes and brought back cover on most of the projects. Also rebased!

**Comment by julienrf** [comment]
Awesome, thanks!

**Pull request merged**